### ISSUE TITLE
[no-useless-constructor] should not report on different constructor type

### ISSUE THREAD

**Issue by JounQin** [typescript-eslint/typescript-eslint#3820]
- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [read the FAQ](https://github.com/typescript-eslint/typescript-eslint/blob/master/docs/getting-started/linting/FAQ.md) and my problem is not listed.

**Repro**

```JSON
{
  "rules": {
    "@typescript-eslint/no-useless-constructor": [2]
  }
}
```

```TS
import { INTERNAL_SERVER_ERROR } from './constants'

import type { ApiError, ApiErrorOptions } from 'shared'

export class ResponseError extends Error implements ApiError {
  declare code: string
  declare reason: string
  declare extra: object | undefined
  declare details: unknown[] | undefined

  constructor(optionOrMsg: Partial<ApiErrorOptions> | string) {
    const options =
      typeof optionOrMsg === 'string' ? { message: optionOrMsg } : optionOrMsg
    super(options.message)
    this.code = options.code || String(INTERNAL_SERVER_ERROR)
    this.reason = options.reason || 'Internal Server Error'
    this.extra = options.extra
    this.details = options.details
  }
}

export class RedirectError extends ResponseError {
  constructor(url: string) { // error reported, unexpected, it changes the constructor type
    super(url)
  }
}
```

**Expected Result**

No error

**Actual Result**

error

**Additional Info**

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `4.29.3` |
| `@typescript-eslint/parser`        | `4.29.3` |
| `TypeScript`                       | `4.4.2` |
| `ESLint`                           | `7.32.0` |
| `node`                             | `12.22.5` |

**Comment by bradzacher** [comment]
> Actual Result
> error

Again, like your last issue, could you please be more specific?
What is the error? What line is it on?

---

I am unable to repro this against master
![image](https://user-images.githubusercontent.com/7462525/131373454-54a3080a-2f25-4bfa-8fee-a2b134f464fa.png)

**Comment by JounQin** [comment]
@bradzacher I've already added comment at `RedirectError`'s constructor, did you scroll the snippet?

**Comment by bradzacher** [comment]
Ah I apologise - I did not see the snippet had more. MacOS hides the scroll bars and the snippet ended perfectly.

This is working as expected - this lint rule does not use type information, only the AST.
And according to the AST - it's a useless constructor as all it does is pass the argument to the `super`.

Either use a disable comment or make the constructor not-useless
```ts
export class RedirectError extends ResponseError {
  constructor(url: string) {
    super({message: url})
  }
}
```

**Comment by JounQin** [comment]
@bradzacher That's OK, but again the limitation could be documented.

**Comment by bradzacher** [comment]
Happy to accept a PR!
If things aren't documented it's usually because it wasn't recognised as a limitation at the time of authoring.
Then so few people run into the limitation that nobody's been motivated to PR it into the docs.

**Comment by JounQin** [comment]
Can you reopen this issue and I'll raise a PR when I'm free, you can assign it to me.

**Comment by JounQin** [comment]
Like https://github.com/typescript-eslint/typescript-eslint/issues/3796#issuecomment-917572367

Can we ignore different constructor names and types? Or at least an option? If you agree, I can try to raise a PR for it/them.

**Comment by bradzacher** [comment]
You cannot do this because you don't have access to type information in this extension rule.
Without type information you cannot inspect the parent's constructor arguments.

**Comment by JounQin** [comment]
@bradzacher Well, can we include type information in this rule then? Just like `unified-signatures`.

It will be a breaking change.

**Comment by bradzacher** [comment]
No.

Significant portions of our userbase does not use type-information. This means that introducing it into an existing rule ***hugely*** restricts the accessibility of the rule. It is a massive breaking change that likely locks users out of using the rule altogether.

So when we make the decision to add type information, it needs to be for a very good reason and add a lot of value.

In addition - adding it into an extension rule is something we need to make sure we're doing for the right reason. Extension rules exist to add support for TS code to the core rules (in this case - accessibility and parameter properties) - so we want them to be as accessible as possible.

Adding type information to handle this (frankly) very rare case is not worth the cost.

**Comment by JoshuaKGoldberg** [comment]
@JounQin are you still planning on sending a PR?

No worries if not, I will soon if you don't have time!

**Comment by JounQin** [comment]
@JoshuaKGoldberg That would be nice! I have no time working on it temporarily.
